admin_app/src/components/DownloadModal.tsx

```diff
@@ -0,0 +1,148 @@
+
+const MAX_CARDS_TO_FETCH = 200;
+
+const DownloadModal = ({
+  open,
+  onClose,
+  onFailedDownload,
+  onNoDataFound,
+}: {
+  open: boolean;
+  onClose: () => void;
+  onFailedDownload: () => void;
+  onNoDataFound: () => void;
+}) => {
+  const { token, accessLevel } = useAuth();
+  const [loading, setLoading] = useState(false);
+
+  const fetchAndCleanContents = async () => {
+    // fetch all contents
+    const raw_json_data = await apiCalls.getContentList({
+      token: token!,
+      skip: 0,
+      limit: MAX_CARDS_TO_FETCH,
+    });
+    if (raw_json_data.length === 0) {
+      return [];
+    }
+    // convert to list of json objects
+    const json_data_list = Object.values(raw_json_data);
+
+    // move content_id to be frst and drop user_id
+    const ordered_json_data_list = json_data_list.map((item) => {
+      const { user_id, content_id, ...rest } = item;
+      return {
+        content_id,
+        ...rest,
+      };
+    });
+    // stringify json element
+    const processed_contents_json = ordered_json_data_list.map((item) => {
+      return {
+        ...item,
+        content_metadata: JSON.stringify(item.content_metadata),
+      };
+    });
+    // get list of tags and replace tag ids with tag names
+    const tags_json = await apiCalls.getTagList(token!);
+    const tag_list = Object.values(tags_json);
+    const tag_dict = tag_list.reduce((acc, tag) => {
+      acc[tag.tag_id] = tag.tag_name;
+      return acc;
+    }, {});
+    processed_contents_json.forEach((item) => {
+      item.content_tags = item.content_tags.map(
+        (tag_id: number) => tag_dict[tag_id],
+      );
+    });
+
+    return processed_contents_json;
+  };
+
+  const downloadCSV = (csvData: string, fileName: string) => {
+    const blob = new Blob([csvData], { type: "text/csv" });
+    const url = window.URL.createObjectURL(blob);
+    const link = document.createElement("a");
+    link.href = url;
+    link.setAttribute("download", fileName);
+    document.body.appendChild(link);
+    link.click();
+    document.body.removeChild(link);
+  };
+
+  const handleDownloadContent = async () => {
+    setLoading(true);
+    try {
+      const processed_contents_json = await fetchAndCleanContents();
+      if (processed_contents_json.length === 0) {
+        onNoDataFound();
+        return;
+      }
+      // convert to csv
+      const csv = Papa.unparse(processed_contents_json);
+      const now = new Date();
+      const timestamp = `${now.getFullYear()}_${String(
+        now.getMonth() + 1,
+      ).padStart(2, "0")}_${String(now.getDate()).padStart(2, "0")}_${String(
+        now.getHours(),
+      ).padStart(2, "0")}${String(now.getMinutes()).padStart(2, "0")}${String(
+        now.getSeconds(),
+      ).padStart(2, "0")}`;
+      const filename = `content_${timestamp}.csv`;
+      downloadCSV(csv, filename);
+    } catch (error) {
+      console.error("Failed to download content", error);
+      onFailedDownload();
```

Comment:
Btw... in general, I would suggest making functions pure - behaviour fully determined by arguments and not by global states and not changing global states. Less of an issue since it's a private function but it makes changing code a lot easier. (Also in Ways of Working)